### ISSUE TITLE
Replace jvm-runner with a non-clojure solution

### ISSUE THREAD

**Issue by jhoffner** [Codewars/codewars-runner-cli#326]
The current clojure based solution is next to impossible for the core team to maintain. We need to move off of this. 

The purpose for the jvm-runner is to improve JVM startup performance. Other existing solutions, such as GroovyServ, should be considered for the replacement.

**Comment by kazk** [comment]
[GroovyServ](https://kobo.github.io/groovyserv/)
- [kobo/groovyserv](https://github.com/kobo/groovyserv)
- [FAQ](https://kobo.github.io/groovyserv/faq.html) has some info about [Drip](https://github.com/ninjudd/drip) and [Nailgun](http://www.martiansoftware.com/nailgun/)

> [Is GroovyServ dedicated for Groovy?](https://kobo.github.io/groovyserv/faq.html#is-groovyserv-dedicated-for-groovy)
> No. you can use GroovyServ for other JVM languages.
> For example, you can run a Clojure script like this:
> ```
> $ CLASSPATH=/tmp/clojure.jar groovyserver
> $ groovyclient -e 'import clojure.main;main.main(args)' -- -e "(println 'Clojure)"
> Clojure
> ```

---

[Scala Native](http://www.scala-native.org/) (pre-release)
- [scala-native/scala-native](https://github.com/scala-native/scala-native)
- [Docs](http://scala-native.readthedocs.io/en/latest/)

**Comment by OverZealous** [comment]
GroovyServ is interesting to me because:

1. Groovy is pretty easy to understand, since it's much closer to traditional C-like languages
2. Groovy has great native Java-library support
3. You can use `@Grab()` to automatically download & include just about any Maven library (and some others), which are cached locally. Obviously that last part only helps for long-running connections, but it could open the door to easily downloading any library upon first run.
4. I believe you can pass in changes to the classpath _after_ warming up the JVM. This has some potential for opening the door to include custom JARs.

-----

I'm not sure how scala-native would help here, as we still need to be running all of our JVM frameworks on the JVM.  This includes Java & Clojure, but also Groovy, Scala, and Kotlin moving forward.

-----

The other option is trying to roll our own, but I think we'll basically end up with something similar to GroovyServ.

**Comment by kazk** [comment]
Added Scala Native because I thought it might be interesting to also consider separating.
If "X Native" existed for all the language supported, this problem can be solved by having `{java,clojure,scala}-runner` instead of `jvm-runner`. I knew about Scala Native and haven't looked into other languages.

**Comment by OverZealous** [comment]
I could see that, though I'd be more concerned about general consistency. Switching to a native runner almost always involves some trade-off, not the least of which is the inability to run JVM libraries.

Obviously Drip and Nailgun are other general purpose solutions to the JVM startup time issues, though the GroovyServ's FAQs (as you pointed out) do a good job of summarizing why they might not work for our needs.

**Comment by kazk** [comment]
Just to be clear, I'm not against GroovyServ at all. It looks good to me and @OverZealous seems to know about it already.

**Comment by OverZealous** [comment]
Obviously the real test will be building out a PoC, which I haven't had time to do in the last … uh … year or so.

**Comment by kazk** [comment]
> We also prewarm containers so that they are ready to be used, which adds to the complexity 

@jhoffner how is this "prewarm" done?

---

I just skimmed through the "User Guide" and the usage will be something like this?

1. prewarm `jvm-runner` container
  a. run `CLASSPATH=/tmp/codewars.jar groovyserver`
  b. "container ready" when `"Server is successfully started up on 1961 port"`
2. on "container ready", do `groovyclient --classpath solution+fixture runTests.groovy`
  a. results in stdout/stderr

**Comment by jhoffner** [comment]
> how is this "prewarm" done?
Great question. I'm so rusty on this I had to look it up and I'm a bit confused myself. The prewarming is done within our server that runs/schedules the CLIs and manages containers. By prewarming its usually just that it has a new container sitting there waiting to be called upon. we use listen.js to keep the container alive. So really its more like pre-starting, not so much pre-warming.

The confusing part is that I thought the jvm-runner would need to actually be pre-warmed. This part of the project wasn't written by me and since i've taken it back over I've half-assed refactored the code around this like 3 times. At one point the jvm runner had a special code path where it was the only container that was pre-started before it was used. When I refactored things I made everything start before it was needed using the listen.js file, and it seems I consolidated the jvm runner to work the same as everything else, bypassing the special JVM code path and utilizing the same resource pool as all of the other languages.

Here is the key part from that file that is no longer being used:

```javascript
        proc = spawn('docker',
                     ['run', '-i', '--rm', '--name', daemon.name, '--entrypoint=/usr/bin/java',
                      container,
                      '-jar', '/codewars/jvm-runner/target/jvm-runner-0.1.1-standalone.jar',
                      readyMessage]);
```

I must be missing something. Java runs a basic request on CW within 3 seconds which is far better than the original 7 or 8 seconds that it used to, so there is clearly still a benefit being received from the jvm runner. This change happened so long ago though that I don't really remember some of the details here.

---------

In terms of prewarming things for the groovyserv (or whatever ends up being used, looks like groovyserv) - we could probably utilize the listen.js script to spawn a prewarmed process if we need to.
